### DISCUSSION TITLE
Question about input X's dimension

### DISCUSSION THREAD

**Discussion by qizzhao** [MilesCranmer/PySR#584] [Q&A] [answered]
I have seen many posts that tutorial about training on 2dimensional's X or 1dimensional's X.  And I am going to make a symbolic regression model, which input X is 3-dimenstion and input y is 2-dimension.  Could `SymbolicRegression.jl`  handle this? If yes, how to do  it. (Please give me a short example please). Thank you.

**Comment by MilesCranmer** [comment, marked as answer]
You can do it with a custom loss function in the `loss_function` parameter: https://astroautomata.com/SymbolicRegression.jl/dev/api/ or https://astroautomata.com/PySR/examples/#9-custom-objectives. The idea is to flatten X into a 2D array, and then reshape it to 3D within the loss function.

To use `eval_tree_array` you have to pass a 2D array. Imagine batching a computation over all pixels.

```julia
const nx = 32
const ny = 20  # For example
function default_objective(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    X = dataset.X
    (prediction, completion) = eval_tree_array(tree, X, options)
    if !completion
        return L(Inf)
    end
    # Here, X is a 2D array of shape [nx * ny, num_features]
    # y is a 1D array of shape [nx * ny]
    
    prediction_reshaped = reshape(prediction, (nx, ny))
    y_reshaped = reshape(dataset.y, (nx, ny))

    # Some computation on the 2D predictions
    
    return L(loss)
end
```

**Reply by qizzhao** [reply]
Thanks for reply. Is there any way that I can set `nx` and `ny` as input parameter of function `default_objective` ?

**Reply by MilesCranmer** [reply]
I’m not sure how they would get passed to it though. But you could just write `nx=32` within the function definition if that’s what you mean?

**Reply by MilesCranmer** [reply]
Alternatively you could store this info in the `extra` field of the `Dataset` class, which permits arbitrary names tuples.

**Reply by qizzhao** [reply]
For my case,  `nx` and `ny` are not constant number, but changing variable number. Each time training would use different value of `nx` and `ny`.  Ideally,  function is like `function default_objective(tree, dataset::Dataset{T,L}, options, nx, ny)` , but it seems not possible.  

How could I save in the `extra` field of the `Dataset` before start of training? ( Do you mean set in `Options` or `EquationSearch` ?)

Thank you.

**Reply by MilesCranmer** [reply]
Why not just do a function closure?

```julia
function build_loss(nx, ny)
    return function default_objective(tree, dataset::Dataset{T,L}, options)::L where {T,L}
        X = dataset.X
        (prediction, completion) = eval_tree_array(tree, X, options)
        if !completion
            return L(Inf)
        end
        # Here, X is a 2D array of shape [nx * ny, num_features]
        # y is a 1D array of shape [nx * ny]
    
        prediction_reshaped = reshape(prediction, (nx, ny))
        y_reshaped = reshape(dataset.y, (nx, ny))

        # Some computation on the 2D predictions
    
        return L(loss)
    end
end
```

then you can pass it to the options like

```julia
Options(loss_function=build_loss(32, 20))
```

so whenever you need to change those numbers, you just change it in that call.

**Reply by qizzhao** [reply]
Thank you.
